src/net_processing.cpp

```diff
@@ -1608,6 +1608,37 @@ void static ProcessGetBlockData(CNode* pfrom, const CChainParams& chainparams, c
     }
 }
 
+//! Determine whether or not a peer can request a transaction, and return it (or nullptr if not found or not allowed).
+CTransactionRef static FindTxForGetData(CNode* peer, const uint256& txid, const std::chrono::seconds mempool_req, const std::chrono::seconds longlived_mempool_time) LOCKS_EXCLUDED(cs_main)
```

Comment:
I'm trying to better understand the lock enforcements & have some questions..

I noticed that `ProcessGetData` has both `LOCKS_EXCLUDED` thread safety annotation as well as an `AssertLockNotHeld` check. I suspect its a product of development over time & is now redundant, but want to confirm my understanding.

`LOCKS_EXCLUDED` enables clang to do static analysis when we compile with the `-Wthread-safety` option. Does the travis build that says `sanitizers: thread (TSan)` run with this option? 

I was trying to understand the limitations of the clang analysis. All I found was a lack of [negative capabilities](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#negative) (aka can't check if a function doesn't have the excluded annotation), which doesn't seem relevant here. 

And then `AssertLockNotHeld()` looks like our custom logic that would do runtime assertions. 

From a kind-of-clumsy search through git history, it looks like `AssertLockNotHeld` was added to `ProcessGetData` before the `LOCKS_EXCLUDED` annotation was. But it looks like `threadsafety.h` has defined `LOCKS_EXCLUDED` since 2014. So I can't color the context of decisions.

To bring this back to this PR, my fundamental question is: what is the difference of guarantees between `LOCKS_EXCLUDED` and `AssertLockNotHeld`? Would there be a reason to add the second check as well?

Reply:
They're somewhat orthogonal, I think.

Annotations:
 * + Is a compile-time check, and guarantee absence of issues in every possible code path
 * - Only works in clang
 * - Can't be used in some more advanced locking scenarios

Assertions:
* + Works in GCC and Clang
* + Isn't restricted to analyzable cases
* - Is only a runtime check; it needs test cases that actually exercise the bug
* - Needs building with -DDEBUG_LOCKORDER

Perhaps at some point it's time to evaluate if we can pick just one over the other (which I guess would be the annotations approach), but they're not exactly identical.

Reply:
ok gotcha. thanks thats very helpful.

so assuming we regularly run the compile-time checks, for this specific case the guarantee from the annotations > guarantee from the assertion. 

I don't see any travis jobs that are running `-Wthread-safety` (and since my previous comment learned that clang thread safety analysis is different than the clang thread sanitizer)

does that mean if somebody added code that acquired `cs_main` and invoked `FindTxForGetData`, we'd be relying on a reviewer to either notice or compile with `-Wthread-safety` to raise an error? 

I do see travis jobs that pass through `-DDEBUG_LOCKORDER`, so assuming we have good functional test coverage, throwing the assertion could cause a travis failure in that circumstance?

It might not be super important / worth invalidating ACKs to add the additional check. The lock expectations have been made very apparent and if new code tried to call this function with the lock acquired, I think we'd catch it in the review process. But on the other hand if we have tooling that reduces dependency on careful review, might as well use it? 

I'm about ready to ACK this PR, just would like to make sure my understanding is clear

Reply:
I think configure.ac turns on `-Wthread-safety-analysis` and `-Werror=thread-safety-analysis` by default if the compiler supports it.

Reply:
ah, thank you.

Reply:
> I don't see any travis jobs that are running `-Wthread-safety` (and since my previous comment learned that clang thread safety analysis is different than the clang thread sanitizer)
> 
> does that mean if somebody added code that acquired `cs_main` and invoked `FindTxForGetData`, we'd be relying on a reviewer to either notice or compile with `-Wthread-safety` to raise an error?

This may be redundant with everything Marco Falke and Jonas Schnelli have set up for the CI and for bitcoinbuilds.org, but I naively reckon it doesn't hurt to independently verify while reviewing. I have a local setup that wipes the slate clean and builds/runs tests with both gcc and clang and all the analysers and sanitizers that can reasonably be run, including `-Wthread-safety`, sometimes on several platforms, that runs in parallel while reviewing non-trivial PRs. It doesn't replace review but does sometimes catch things.